cmake-modules/add_nlohmann_json.cmake

```diff
@@ -0,0 +1,30 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# SPDX-License-Identifier: MIT
+#
+# Defines utility functions to create code coverage targets with gcov.
+# gcov html and xml report.
+#
+
+# Azure Core can expose the Nlohmann Json header by fetching a release version and including the header to the include path.
+# 
+# Azure Core will try to find the json header first. This is for an application that is already using Nlohmann header.
+# When it is not found it, Azure Core will featch it.
+
+# Storage requires 3.8 version for using `contains` feature
+find_package(nlohmann_json 3.8.0 CONFIG)
+if (NOT nlohmann_json_FOUND)
+
+    include(FetchContent)
+
+    # realese code only. This save us from getting the entire nlohmann source code.
+    FetchContent_Declare(json
```

Comment:
Why do we download rather than install via vcpkg like other dependencies?

Reply:
well, one of the benefits of this json lib is that it is just a header file with no src. All we need is to download the json header and included as an interface.

The end user can still get it with vcpkg or install it in the system, in that case, we would find it with `find_package` and won't download anything.

What I didn't like is to block the compilation for a missing lib that is only a header file that we can easily get. There is no multiple compilation flavors for it (like we have for other libs like libcurl) where the end user might want to get and compile the dependency with some specific config.
So, trying to simplify the customer experience.

And one last thing, is that this is still part of a temporal functionality that is not yet certified and accepted for Azure Core to GA, so, we will most likely change it

Reply:
Sure, but in this way, you always get a specific version, an old version.

Reply:
Even if we used vcpkg, we'd get a specific pinned version (because we use a specific release of vcpkg), so that aspect remains identical to what @vhvb1989 is doing here. And that's also identical to having a copy of a particular version checked-in.

The version upgrades, like all other dependencies, will need to be intentional.

Reply:
But as new versions are released in vcpkg, people can get newer versions, we don't need to do anything.

But since you've written a specific version 3.8.0 in our CMake file https://github.com/Azure/azure-sdk-for-cpp/blob/71d6338d9c91dfc7ef3ee3a0eef335cfafecfa1f/cmake-modules/add_nlohmann_json.cmake#L22. people can only get this specific version, unless we bump up this version, which will increase our maintenance cost.

Reply:
Yeah, keep in mind this is not the final decision. Azure Core will still go further about the entire Json story and elaborate about the upgrading strategy and maintenance.

Reply:
> But as new versions are released in vcpkg, people can get newer versions, we don't need to do anything.

> But since you've written a specific version 3.8.0 in our CMake file

If that is an important scenario to enable, I believe the `CONFIG` setting in `find_package` will match the newer version that is on the user's machine, and we won't be going down this path to download the 3.8.0 version. This path only runs if `find_package` fails.